language/move-prover/bytecode-to-boogie/src/prelude.bpl

```diff
@@ -0,0 +1,543 @@
+// ================================================================================
+// Domains
+
+type TypeName;
+type FieldName = int;
+type LocalName;
+type Address = int;
+type ByteArray;
+type String;
+type {:datatype} Edge;
+function {:constructor} Field(f: FieldName): Edge;
+function {:constructor} Index(i: int): Edge;
+function {:constructor} String(s: String): Edge;
+
+type {:datatype} Path;
+function {:constructor} Path(p: [int]Edge, size: int): Path;
+const EmptyPath: Path;
+axiom size#Path(EmptyPath) == 0;
+
+type {:datatype} TypeValue;
+function {:constructor} BooleanType() : TypeValue;
+function {:constructor} IntegerType() : TypeValue;
+function {:constructor} AddressType() : TypeValue;
+function {:constructor} ByteArrayType() : TypeValue;
+function {:constructor} StrType() : TypeValue;
+function {:constructor} VectorType(t: TypeValue) : TypeValue;
+function {:constructor} StructType(name: TypeName, ts: TypeValueArray) : TypeValue;
+const DefaultTypeValue: TypeValue;
+function {:builtin "MapConst"} MapConstTypeValue(tv: TypeValue): [int]TypeValue;
+
+type {:datatype} TypeValueArray;
+function {:constructor} TypeValueArray(v: [int]TypeValue, l: int): TypeValueArray;
+const EmptyTypeValueArray: TypeValueArray;
+axiom l#TypeValueArray(EmptyTypeValueArray) == 0;
+axiom v#TypeValueArray(EmptyTypeValueArray) == MapConstTypeValue(DefaultTypeValue);
+
+type {:datatype} Value;
+function {:constructor} Boolean(b: bool): Value;
+function {:constructor} Integer(i: int): Value;
+function {:constructor} Address(a: Address): Value;
+function {:constructor} ByteArray(b: ByteArray): Value;
+function {:constructor} Str(a: String): Value;
+function {:constructor} Vector(v: ValueArray): Value;
+const DefaultValue: Value;
+function {:builtin "MapConst"} MapConstValue(v: Value): [int]Value;
```

Comment:
What is `{:builtin "MapConst"}` doing here?

Reply:
This is a Boogie feature to target native theories.  The :builtin attribute is tracked by the part of Boogie that converts code to verification conditions and if it recognizes the string attached to :builtin, it can treat it specially.  In this case, we are telling Boogie to use the constant map constructor provided by Z3.